File: web/status-history/src/main/java/net/twisterrob/travel/statushistory/view/handlebars/HandlebarsHelpers.java
package net.twisterrob.travel.statushistory.view.handlebars;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.EnumMap;
import java.util.Locale;

import com.github.jknack.handlebars.Options;

public class HandlebarsHelpers {

	public static void assign(String varName, Object varValue, Options options) {
		options.data(varName, varValue);
	}

	public static int add(int base, int increment) {
		return base + increment;
	}

	public static String concat(String base, String extra) {
		return base + extra;
	}

	public static boolean not(boolean value) {
		return !value;
	}

	public static boolean or(boolean left, boolean right) {
		return left || right;
	}

	public static boolean and(boolean left, boolean right) {
		return left && right;
	}

	public static boolean empty(String value) {
		return value == null || value.isEmpty();
	}

	public static String formatDate(Date value, String pattern) {
		return new SimpleDateFormat(pattern, Locale.getDefault()).format(value);
	}

	/**
	 * Workaround for `myEnumMap.[Foo]` and `lookup myEnumMap Foo` not working.
	 * Usage: Replace {@code (lookup someEnumMap someEnumKey)} with {@code (lookupEnumMap someEnumMap someEnumKey)}.
	 * @see https://github.com/jknack/handlebars.java/issues/1083
	 */
	public static <E extends Enum<E>> Object lookupEnumMap(EnumMap<E, ?> map, E key) {
		return map.get(key);
	}
}

File: web/status-history/src/main/java/net/twisterrob/travel/statushistory/view/handlebars/HandlebarsHelpers.kt
package net.twisterrob.travel.statushistory.view.handlebars

import com.github.jknack.handlebars.Options
import net.twisterrob.java.exceptions.StackTrace
import org.slf4j.LoggerFactory
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

@Suppress("unused") // Used by .hbs files.
object HandlebarsHelpers {

	@JvmStatic
	fun debug(obj: Any?): String {
		val value = if (obj != null) "(${obj::class.java}) ${obj}" else "<null>"
		LoggerFactory.getLogger(HandlebarsHelpers::class.java).debug(value, StackTrace())
		return value
	}

	@JvmStatic
	fun assign(varName: String, varValue: Any?, options: Options) =
		options.data(varName, varValue)

	@JvmStatic
	fun add(base: Int, increment: Int): Int =
		base + increment

	@JvmStatic
	fun concat(base: String, extra: String): String =
		base + extra

	@JvmStatic
	fun not(value: Boolean): Boolean =
		!value

	@JvmStatic
	fun or(left: Boolean, right: Boolean): Boolean =
		left || right

	@JvmStatic
	fun and(left: Boolean, right: Boolean): Boolean =
		left && right

	@JvmStatic
	fun empty(value: String?): Boolean =
		value.isNullOrEmpty()

	@JvmStatic
	fun formatDate(value: Date, pattern: String): String =
		SimpleDateFormat(pattern, Locale.getDefault()).format(value)

	@JvmStatic
	fun formatCssColor(value: Int): String =
		"#%06X".format(Locale.ROOT, value and @Suppress("detekt.MagicNumber") 0xFFFFFF)
}
